### ISSUE TITLE
Unnecessary copy of ACISM struct?

### ISSUE THREAD

**Issue by BabeleDunnit** [mischasan/aho-corasick#18]
Hi,

I am using your (very good) library to do character-by-character AC matching, i.e. I am calling acism_more() passing a "sliding window" of length 1 on a const char*. I noticed this:

https://github.com/mischasan/aho-corasick/blob/master/acism.c#L29

this seems avoidable, and in a use-case like mine, removing that and just using the ACISM\* passed as parameter (i.e. "psp") in the rest of the acism_more() code leads to a 45% speed boost (!).

Am I missing something? :)

ciao!
Aaron

**Comment by mischasan** [comment]
Copying *psp was a case of premature optimization :-) trying to save the
compiler one more register.
You're not missing anything, and it certainly makes sense in your use case.
I think I'll just flip it to work that way.
Thanks again. And you're the first person I've heard of who's using
acism_more that way. What (in general terms) is the application?

**Comment by BabeleDunnit** [comment]
Hi Mischa,

:) "premature optimization is the root of all evil" bla bla bla :)

With modern multicore and cache processor architectures, optimizing has become very hard. Not so strange that code you would swear to be faster turn out to be slower... So you have all my simpathy :)

I am developing a library which "recognizes strings" :) . I am investigating several algorithms. One of them (the best one so far) needs to do a search and replace on the fly of known substring into one-byte tokens while trying to match the (possibly very long) string vs. a (possibly very big) set of reference (tokenized) strings. Until now, the one-char-at-a-time approach is the best one because has a very low failure time (I mean, "I don't know this string, exit"), so other match engines can be used if this one fails.

Using your library was initially disappointing: I made some preliminary test, and anyway it is enough to look at the code, nothing can be faster than accessing tables. It HAD to be as fast as hell. But in my case the whole thing resulted to be slower than before (I was using Multifast lib, because for some fu@&&ng reason I did not find ACISM before. I would have used it from day one), and I felt miserable for fifteen minutes. Then I rebooted my brain and tracked down the poorer performance to the obvious fact that in my use case acism_more() is called much more times than you would in a "normal" way. The remaining part was easy.

It seems incredible but avoiding that copy made my code 2.5 / 3 times faster, so thank you! The bottleneck is not the AC any more!

Later,

Aaron

> Il giorno 05 ott 2016, alle ore 21:02, Mischa Sandberg <notifications@github.com> ha scritto:
> 
> Copying *psp was a case of premature optimization :-) trying to save the
> compiler one more register.
> You're not missing anything, and it certainly makes sense in your use case.
> I think I'll just flip it to work that way.
> Thanks again. And you're the first person I've heard of who's using
> acism_more that way. What (in general terms) is the application?
>
